bcgov/cas-registration
# App static layout structure (plus...)

Addresses [ticket 175](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/175)

## 🚧 Changes

✨ Components:
    - Add client\middleware.ts for request routing management
    - Add client\middlewares\ * files for task specific functions applied to a request (TBD)
    - Add client\app\auth\signin\page.tsx for login route (TBD)
    - Add client\app\components\theme\* for MUI custom theme with Next.js 13 app router
    - Modify client\app\layout.tsx to use ThemeRegistry
    - Modify client\next.config.js for MUI imports
    - Add client\app\components\layout\Header for static app layout
    - Add client\app\components\layout\Footer for static app layout
    
🧹 Cleanup:
    - Remove Tailwindcss configurations files
    - Modify package.json to resolve dependancy warnings
    - Remove client\app\styles\global.css
    - Remove client\app\styles\layout.module.css    
    - Rename client\app\utils\actions\operationSubmitHandler.ts to utils\actions.ts, as per Next.js convention
    - Add client\.env\API_URL variable for API requests URL
    - Add client\app\utils\api.ts fetch utility function to handle API requests and error catching
    - Re-code fetch calls to hardcoded API to use utils\api.ts fetch utility function, as per DRY convention
    - remove client\app\components\loading.tsx
    
🏗️ Redesign:
    - Delete client\app\components\loading.tsx for suspend loading effect
    - Add client\app\components\loading\SkeletonGrid.tsx for suspend loading effect
    - Add client\app\components\loading\SkeletonForm.tsx for suspend loading effect
    - Add client\app\components\routes\operations\Operations.tsx (contains async components from client\app\operations\page.tsx)
    - Recode client\app\operations\page.tsx to wrap asynchronous client\app\components\operations\Operations.tsx using <Suspense>
    - Add client\app\components\routes\operations\form\Operation.tsx (contains async components from client\app\operations\create\page.tsx)
    - Recode client\app\operations\create\page.tsx to wrap asynchronous client\app\components\operations\form\Operation.tsx using <Suspense> 
    

 ## 🚀 Impact:
   - Material UI configured with the Next.js App Router
   - Initial static layout structure (**rudimentary and non-functional- refinement TBD**)
   -  Middleware request routing architecture in place for future requirements of request rewriting, redirecting, or modifying before a response
   - Re-architecture of app router files for wrapping asynchronous components in page routes to improved user interaction. 
**Ex:** 
Route http://localhost:3000/operations displays page.tsx which displays none dynamic elements and uses <Suspense> with a loading skeleton to indicate waiting for the async data components

## 📝  Notes:
- Added a temp latency simulation on async data to emphasize the user interaction with <Suspense> skeleton loading


##  🔬 Testing:

1. Open terminal in cas-registration\bc_obps directory
2. Follow the [developer environment setup](https://github.com/bcgov/cas-registration/blob/develop/docs/developer-environment-setup.md#backend-environment-setup) to start the API
3. Open a terminal in cas-registration\client folder
4. Install changes to the packgae.json
   ```
   yarn
   ```
5. Start the development server:
   ```
   yarn dev
   ```
6. In your web browser, navigate to `http://localhost:3000/`
7. You should see "navigation" page with MUI header and footer
![image](https://github.com/bcgov/cas-registration/assets/120038448/e238d7ce-5f05-4818-a798-169c2ab271f3)
8. Click link "Operations"
9. You should see "Operations List" page with loading skeleton
![image](https://github.com/bcgov/cas-registration/assets/120038448/5186a5cf-b27c-42f0-850d-b559235d5977)
10. Once loading has completed you should see list of operations
11. Click button "Add Operation"
12. You should see "Create a New Operation" page with loading skeleton
![image](https://github.com/bcgov/cas-registration/assets/120038448/fc27680e-62e8-4ee2-a4fd-a0235d9238bf)
13. Once loading is complete you should see an empty form
14. Complete and submit form
15. A success message should display
16. Click the back button
17. New record should display
18. Click the "View Details" on a row
19.. You should see "Edit an Operation" page with loading skeleton
![image](https://github.com/bcgov/cas-registration/assets/120038448/89bcf029-46f2-456b-985a-5db859b5b867)
20.  Once loading is complete you should see an form with data from the selected row.
21. Edit submit form
22. A success message should display
23. Click the back button
24. Edits should display
